posts/virtual_environments.md

```diff
@@ -0,0 +1,126 @@
+- The data hash of the model is based on model attributes that have a direct impact on the dataset produced by this model. Such attributes include the model's query, storage format, partitioning scheme, etc. When this hash changes the model is considered to be modified **directly**.
+- The data hash of upstream models. This is the same data hash calculation logic but applied to models that are upstream from the current model in order to capture data changes caused by the model's dependencies. Similarly, when this hash changes the model is considered to be modified **indirectly**.
+- The metadata hash of the model is based on model attributes that have no impact on data itself. This includes metadata like ownership information, descriptions, comments, etc.
+- The metadata hash of upstream models captures metadata changes in model's dependencies.
+
+Additionally, thanks to the [SQLGlot optimizer](https://github.com/tobymao/sqlglot/blob/main/posts/python_sql_engine.md#optimizing) we’re able to create a canonical version of each SQL query, so that non-functional changes like formatting or structural refactoring don’t impact the fingerprint.
+
+#### Automatic Change Categorization
+The difference between **direct** and **indirect** modifications becomes relevant when [categorizing](https://sqlmesh.readthedocs.io/en/stable/concepts/plans/#change-categories) individual changes during plan creation.
+
+Every time a model is modified directly, SQLMesh automatically categorizes the change as “breaking” or “non-breaking” based on the change’s impact on models that are downstream from the modified one. This is possible thanks to the platform’s ability to understand SQL semantically, at the [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) level. By comparing the current model version with the previous one we’re able to compute a [semantic diff](https://github.com/tobymao/sqlglot/blob/main/posts/sql_diff.md) which SQLMesh further analyzes to understand what kind of impact it would have on downstream models.
+
+Thus, for example, adding a new column is not considered to be a breaking change but removing or modifying the existing one - is.
+
+![Figure 4: Breaking vs. non-breaking changes](virtual_environments/change_categorization.png)
+*Figure 4: Breaking vs. non-breaking changes.*
+
+This way SQLMesh can automatically reduce the amount of necessary recomputation to a minimum while guaranteeing correctness. No user intervention needed.
+
+Presently this approach doesn’t extend beyond looking at each model as a whole. Soon, however, we’ll be able to categorize changes per individual column instead, thanks to the platform’s support for column-level lineage. The platform will achieve an even finer balance between correctness and efficiency, since changes like removing a column that is not referenced downstream will no longer be categorized as a “breaking”.
```

Comment:
Presently changes are checked at the model level; however, we'll soon be extending this to the individual column level!